### PR TITLE
19421 - Qualified Supplier Product resubmission

### PR DESCRIPTION
**Pull request by ochiu**
*Issue #:*
https://github.com/bcgov/entity/issues/19421

*Description of changes:*
- PATCH endpoint on org products for resubmitting a product subscription request
- model change for product_codes indicating which products allow resubmission
- model change for task to capture that a task has be resubmitted for the re-approved flow (qualified supplier)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-auth license (Apache 2.0).

### CODE REVIEW

File: auth-api/src/auth_api/resources/v1/org_products.py [hunk 1]
```diff
@@ -65,3 +65,22 @@ def post_org_product_subscription(org_id):
     except BusinessException as exception:
         response, status = {'code': exception.code, 'message': exception.message}, exception.status_code
     return response, status
+
+
+@bp.route('', methods=['PATCH'])
+@cross_origin(origins='*')
+@TRACER.trace()
+@_jwt.has_one_of_roles([Role.PUBLIC_USER.value])
```
Comment (seeker25) [inline auth-api/src/auth_api/resources/v1/org_products.py:73]: Do staff need this functionality too? They don't have `PUBLIC_USER`

File: auth-api/src/auth_api/services/products.py [hunk 2]
```diff
@@ -76,6 +77,63 @@ def find_product_type_by_code(code: str) -> str:
         product_code_model: ProductCodeModel = ProductCodeModel.find_by_code(code)
         return getattr(product_code_model, 'type_code', '')
 
+    @staticmethod
+    def _validate_product_resubmission(task: TaskModel, product_model: ProductCodeModel):
+        # Check that the product code can be resubmitted and requires review
```
Comment (seeker25) [inline auth-api/src/auth_api/services/products.py:82]: suggestion: redundant, the code explains that below

File: auth-api/src/auth_api/services/products.py [hunk 3]
```diff
@@ -76,6 +77,63 @@ def find_product_type_by_code(code: str) -> str:
...
+
+        A re-submission is only possible when:
+        1) product_code.can_resubmit flag is True
+        2) product_code.need_review flag is True
+        3) Task is currently in a REJECTED state.
+        """
+        org: OrgModel = OrgModel.find_by_org_id(org_id)
+        if not org:
+            raise BusinessException(Error.DATA_NOT_FOUND, None)
+        # Check authorization for the user
```
Comment (seeker25) [inline auth-api/src/auth_api/services/products.py:101]: suggestion: redundant, the code explains that below

File: auth-api/src/auth_api/services/products.py [hunk 4]
```diff
@@ -76,6 +77,63 @@ def find_product_type_by_code(code: str) -> str:
...
+            product_model: ProductCodeModel = ProductCodeModel.find_by_code(product_code)
+
+            # We only care about existing subs for resubmission
+            if not existing_sub:
+                continue
+
+            task: TaskModel = TaskModel.find_by_task_relationship_id(existing_sub.id,
+                                                                     TaskRelationshipType.PRODUCT.value,
+                                                                     TaskStatus.COMPLETED.value)
+            # Check if the review task is valid for resubmission
```
Comment (seeker25) [inline auth-api/src/auth_api/services/products.py:119]: suggestion: redundant, the code explains that below

File: auth-api/src/auth_api/resources/v1/org_products.py (hunk 1 above)
Comment (ochiu) [reply auth-api/src/auth_api/resources/v1/org_products.py:73]: Staff will be using the dashboard to transition the status of a task. This is meant to be be triggered from a public user when they resubmit an application specifically Qualified Supplier in this case

File: auth-api/src/auth_api/services/products.py [hunk 5]
```diff
@@ -76,6 +77,63 @@ def find_product_type_by_code(code: str) -> str:
...
+            Product._validate_product_resubmission(task, product_model)
+
+            # Reset the state of the review task
```
Comment (seeker25) [inline auth-api/src/auth_api/services/products.py:122]: suggestion: redundant, the code explains that below

File: auth-api/src/auth_api/services/products.py [hunk 6]
```diff
@@ -76,6 +77,63 @@ def find_product_type_by_code(code: str) -> str:
...
+            Product._reset_subscription_and_review_task(review_task=task,
+                                                        product_model=product_model,
+                                                        subscription=existing_sub,
+                                                        user_id=user.id)
+
+            # Resend confirmations
```
Comment (seeker25) [inline auth-api/src/auth_api/services/products.py:128]: suggestion: redundant, the code explains that below

**Review by seeker25 — APPROVED** [review]
LGTM

File: auth-api/src/auth_api/resources/v1/org_products.py (hunk 1 above)
Comment (seeker25) [reply auth-api/src/auth_api/resources/v1/org_products.py:73]: auth check in there too, good stuff

**Pull request merged**